File: bench/java/src/test/java/ming/ConcurrencyTest.java
package ming;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Assumptions;
import java.util.List;
import java.util.ArrayList;
import java.util.concurrent.*;
import static org.junit.jupiter.api.Assertions.*;

/**
 * Level 27: Concurrent Evaluation
 * eval_str must be safe for concurrent use from multiple threads.
 */
class ConcurrencyTest {

    private static int benchLevel() {
        String env = System.getenv("BENCH_LEVEL");
        if (env == null) env = System.getProperty("bench.level", "0");
        try { return Integer.parseInt(env); } catch (NumberFormatException e) { return 0; }
    }

    private void skipIfBelowLevel() {
        int level = benchLevel();
        Assumptions.assumeTrue(level == 0 || level >= 27, "Skipping: BENCH_LEVEL < 27");
    }

    @Test
    void testL27ConcurrentIndependentEval() throws Exception {
        skipIfBelowLevel();
        ExecutorService pool = Executors.newFixedThreadPool(8);
        List<Future<String>> futures = new ArrayList<>();

        for (int i = 0; i < 8; i++) {
            final int seed = i;
            futures.add(pool.submit(() -> {
                Interpreter interp = new Interpreter();
                return interp.evalStr(String.format(
                    "(let loop ((n 1000) (acc 0)) (if (= n 0) acc (loop (- n 1) (+ acc %d))))", seed));
            }));
        }

        for (int i = 0; i < 8; i++) {
            assertEquals(String.valueOf(i * 1000), futures.get(i).get(30, TimeUnit.SECONDS));
        }
        pool.shutdown();
    }

    @Test
    void testL27ConcurrentOutputIsolation() throws Exception {
        skipIfBelowLevel();
        ExecutorService pool = Executors.newFixedThreadPool(4);
        List<Future<Interpreter.EvalResult>> futures = new ArrayList<>();

        for (int i = 0; i < 4; i++) {
            final int idx = i;
            futures.add(pool.submit(() -> {
                Interpreter interp = new Interpreter();
                return interp.evalStrWithOutput(String.format(
                    "(begin (display \"thread%d\") (display \" \") (display \"done%d\") \"ok\")", idx, idx));
            }));
        }

        for (int i = 0; i < 4; i++) {
            Interpreter.EvalResult r = futures.get(i).get(30, TimeUnit.SECONDS);
            assertEquals("ok", r.result());
            assertEquals(String.format("thread%d done%d", i, i), r.output());
        }
        pool.shutdown();
    }

    @Test
    void testL27ConcurrentClosuresAndMutation() throws Exception {
        skipIfBelowLevel();
        ExecutorService pool = Executors.newFixedThreadPool(4);
        List<Future<String>> futures = new ArrayList<>();

        for (int i = 0; i < 4; i++) {
            futures.add(pool.submit(() -> {
                Interpreter interp = new Interpreter();
                return interp.evalStr(
                    "(let ((count 0))" +
                    "  (define (inc!) (set! count (+ count 1)) count)" +
                    "  (inc!) (inc!) (inc!)" +
                    "  count)");
            }));
        }

        for (Future<String> f : futures) {
            assertEquals("3", f.get(30, TimeUnit.SECONDS));
        }
        pool.shutdown();
    }

    @Test
    void testL27ConcurrentStress() throws Exception {
        skipIfBelowLevel();
        ExecutorService pool = Executors.newFixedThreadPool(16);
        List<Future<String>> futures = new ArrayList<>();

        for (int i = 0; i < 16; i++) {
            final int idx = i;
            futures.add(pool.submit(() -> {
                Interpreter interp = new Interpreter();
                return interp.evalStr(String.format(
                    "(let ((x %d)) (define (f n) (if (= n 0) x (f (- n 1)))) (f 100))", idx));
            }));
        }

        for (int i = 0; i < 16; i++) {
            assertEquals(String.valueOf(i), futures.get(i).get(30, TimeUnit.SECONDS));
        }
        pool.shutdown();
    }

    // ===== State isolation tests (sequential) =====

    @Test
    void testL27SequentialStateLeak() throws Exception {
        skipIfBelowLevel();
        Interpreter interp = new Interpreter();
        String r1 = interp.evalStr("(begin (define x 42) x)");
        assertEquals("42", r1);

        // x must not leak to next eval_str call
        Interpreter interp2 = new Interpreter();
        assertThrows(EvalError.class, () -> interp2.evalStr("x"),
            "variable 'x' leaked between independent evalStr calls");
    }

    @Test
    void testL27SequentialOutputLeak() throws Exception {
        skipIfBelowLevel();
        Interpreter i1 = new Interpreter();
        Interpreter.EvalResult r1 = i1.evalStrWithOutput("(display \"aaa\")");
        Interpreter i2 = new Interpreter();
        Interpreter.EvalResult r2 = i2.evalStrWithOutput("(display \"bbb\")");
        assertEquals("aaa", r1.output());
        assertEquals("bbb", r2.output(), "output buffer leaked between evalStrWithOutput calls");
    }

    // ===== Concurrent continuation/macro collision tests =====

    @Test
    void testL27ConcurrentCallccCollision() throws Exception {
        skipIfBelowLevel();
        ExecutorService pool = Executors.newFixedThreadPool(8);
        List<Future<String>> futures = new ArrayList<>();

        for (int i = 0; i < 8; i++) {
            futures.add(pool.submit(() -> {
                Interpreter interp = new Interpreter();
                return interp.evalStr(
                    "(let ((count 0))" +
                    "  (set! count (+ count (call/cc (lambda (k) (k 10)))))" +
                    "  count)");
            }));
        }

        for (Future<String> f : futures) {
            assertEquals("10", f.get(30, TimeUnit.SECONDS));
        }
        pool.shutdown();
    }

    @Test
    void testL27ConcurrentMacroHygiene() throws Exception {
        skipIfBelowLevel();
        ExecutorService pool = Executors.newFixedThreadPool(4);
        List<Future<String>> futures = new ArrayList<>();

        for (int i = 0; i < 4; i++) {
            futures.add(pool.submit(() -> {
                Interpreter interp = new Interpreter();
                return interp.evalStr(
                    "(begin" +
                    "  (define-syntax my-swap!" +
                    "    (syntax-rules ()" +
                    "      ((_ a b) (let ((tmp a)) (set! a b) (set! b tmp)))))" +
                    "  (let ((x 1) (y 2))" +
                    "    (my-swap! x y)" +
                    "    (list x y)))");
            }));
        }

        for (Future<String> f : futures) {
            assertEquals("(2 1)", f.get(30, TimeUnit.SECONDS));
        }
        pool.shutdown();
    }
}

File: bench/hidden/go/concurrency_test.go
package ming

import (
	"fmt"
	"testing"
)

// ===== Level 27: Concurrent Evaluation =====
// eval_str must be safe for concurrent use from multiple goroutines.

func skipIfBelowL27(t *testing.T) {
	t.Helper()
	level := benchLevel()
	if level > 0 && level < 27 {
		t.Skip("Skipping: BENCH_LEVEL < 27")
	}
}

func TestL27ConcurrentIndependentEval(t *testing.T) {
	skipIfBelowL27(t)
	const n = 8
	type result struct {
		idx int
		val string
		err error
	}
	ch := make(chan result, n)
	for i := 0; i < n; i++ {
		go func(seed int) {
			program := fmt.Sprintf("(let loop ((n 1000) (acc 0)) (if (= n 0) acc (loop (- n 1) (+ acc %d))))", seed)
			val, err := EvalStr(program)
			ch <- result{seed, val, err}
		}(i)
	}
	for i := 0; i < n; i++ {
		r := <-ch
		if r.err != nil {
			t.Fatalf("thread %d: unexpected error: %v", r.idx, r.err)
		}
		expected := fmt.Sprintf("%d", r.idx*1000)
		if r.val != expected {
			t.Fatalf("thread %d: expected %q, got %q", r.idx, expected, r.val)
		}
	}
}

func TestL27ConcurrentOutputIsolation(t *testing.T) {
	skipIfBelowL27(t)
	const n = 4
	type result struct {
		idx    int
		val    string
		output string
		err    error
	}
	ch := make(chan result, n)
	for i := 0; i < n; i++ {
		go func(idx int) {
			program := fmt.Sprintf(`(begin (display "thread%d") (display " ") (display "done%d") "ok")`, idx, idx)
			val, output, err := EvalStrWithOutput(program)
			ch <- result{idx, val, output, err}
		}(i)
	}
	for i := 0; i < n; i++ {
		r := <-ch
		if r.err != nil {
			t.Fatalf("thread %d: unexpected error: %v", r.idx, r.err)
		}
		if r.val != "ok" {
			t.Fatalf("thread %d: expected \"ok\", got %q", r.idx, r.val)
		}
		expectedOutput := fmt.Sprintf("thread%d done%d", r.idx, r.idx)
		if r.output != expectedOutput {
			t.Fatalf("thread %d: expected output %q, got %q", r.idx, expectedOutput, r.output)
		}
	}
}

func TestL27ConcurrentClosuresAndMutation(t *testing.T) {
	skipIfBelowL27(t)
	const n = 4
	ch := make(chan struct {
		val string
		err error
	}, n)
	for i := 0; i < n; i++ {
		go func() {
			val, err := EvalStr(`(let ((count 0))
				(define (inc!) (set! count (+ count 1)) count)
				(inc!) (inc!) (inc!)
				count)`)
			ch <- struct {
				val string
				err error
			}{val, err}
		}()
	}
	for i := 0; i < n; i++ {
		r := <-ch
		if r.err != nil {
			t.Fatalf("unexpected error: %v", r.err)
		}
		if r.val != "3" {
			t.Fatalf("expected \"3\", got %q", r.val)
		}
	}
}

func TestL27ConcurrentStress(t *testing.T) {
	skipIfBelowL27(t)
	const n = 16
	type result struct {
		idx int
		val string
		err error
	}
	ch := make(chan result, n)
	for i := 0; i < n; i++ {
		go func(idx int) {
			program := fmt.Sprintf("(let ((x %d)) (define (f n) (if (= n 0) x (f (- n 1)))) (f 100))", idx)
			val, err := EvalStr(program)
			ch <- result{idx, val, err}
		}(i)
	}
	for i := 0; i < n; i++ {
		r := <-ch
		if r.err != nil {
			t.Fatalf("thread %d: unexpected error: %v", r.idx, r.err)
		}
		expected := fmt.Sprintf("%d", r.idx)
		if r.val != expected {
			t.Fatalf("thread %d: expected %q, got %q", r.idx, expected, r.val)
		}
	}
}

// ===== State isolation tests (sequential) =====

func TestL27SequentialStateLeak(t *testing.T) {
	skipIfBelowL27(t)
	r1, err := EvalStr("(begin (define x 42) x)")
	if err != nil {
		t.Fatalf("first eval failed: %v", err)
	}
	if r1 != "42" {
		t.Fatalf("expected \"42\", got %q", r1)
	}
	_, err2 := EvalStr("x")
	if err2 == nil {
		t.Fatal("variable 'x' leaked between independent EvalStr calls")
	}
}

func TestL27SequentialOutputLeak(t *testing.T) {
	skipIfBelowL27(t)
	_, out1, err1 := EvalStrWithOutput(`(display "aaa")`)
	if err1 != nil {
		t.Fatalf("first eval failed: %v", err1)
	}
	_, out2, err2 := EvalStrWithOutput(`(display "bbb")`)
	if err2 != nil {
		t.Fatalf("second eval failed: %v", err2)
	}
	if out1 != "aaa" {
		t.Fatalf("expected output \"aaa\", got %q", out1)
	}
	if out2 != "bbb" {
		t.Fatalf("output buffer leaked: expected \"bbb\", got %q", out2)
	}
}

func TestL27ConcurrentCallccCollision(t *testing.T) {
	skipIfBelowL27(t)
	const n = 8
	ch := make(chan struct {
		val string
		err error
	}, n)
	for i := 0; i < n; i++ {
		go func() {
			val, err := EvalStr(`(let ((count 0))
				(set! count (+ count (call/cc (lambda (k) (k 10)))))
				count)`)
			ch <- struct {
				val string
				err error
			}{val, err}
		}()
	}
	for i := 0; i < n; i++ {
		r := <-ch
		if r.err != nil {
			t.Fatalf("unexpected error: %v", r.err)
		}
		if r.val != "10" {
			t.Fatalf("expected \"10\", got %q", r.val)
		}
	}
}

func TestL27ConcurrentMacroHygiene(t *testing.T) {
	skipIfBelowL27(t)
	const n = 4
	ch := make(chan struct {
		val string
		err error
	}, n)
	for i := 0; i < n; i++ {
		go func() {
			val, err := EvalStr(`(begin
				(define-syntax my-swap!
					(syntax-rules ()
						((_ a b) (let ((tmp a)) (set! a b) (set! b tmp)))))
				(let ((x 1) (y 2))
					(my-swap! x y)
					(list x y)))`)
			ch <- struct {
				val string
				err error
			}{val, err}
		}()
	}
	for i := 0; i < n; i++ {
		r := <-ch
		if r.err != nil {
			t.Fatalf("unexpected error: %v", r.err)
		}
		if r.val != "(2 1)" {
			t.Fatalf("expected \"(2 1)\", got %q", r.val)
		}
	}
}
